# ML lab 01 - TensorFlow의 설치및 기본적인 operations

In [1]:
import tensorflow as tf

/Users/angelx/.pyenv/versions/3.6.1/lib/python3.6/importlib/_bootstrap.py:205: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)


In [2]:
tf.__version__

'1.10.1'

## Hello TensorFlow!

- create a constant op
- this op is added as a node to the default graph

In [3]:
hello = tf.constant("Hello TensorFlow!")

In [4]:
hello

<tf.Tensor 'Const:0' shape=() dtype=string>

- start a TF session

In [5]:
sess = tf.Session()

- run the op and get result

In [6]:
print(sess.run(hello))

b'Hello TensorFlow!'


## Computational Graph

- Build graph (tensors) using TensorFlow operations

In [7]:
node1 = tf.constant(3.0, tf.float32)
node2 = tf.constant(4.0) # also tf.float32 implicitly
node3 = tf.add(node1, node2)

In [8]:
print('node1:', node1, 'node2:', node2)
print('node3:', node3)

node1: Tensor("Const_1:0", shape=(), dtype=float32) node2: Tensor("Const_2:0", shape=(), dtype=float32)
node3: Tensor("Add:0", shape=(), dtype=float32)


- feed data and run graph(operation) sess.run(op)
- update varialbes in the graph(and return values)

In [9]:
sess = tf.Session()
print('sess.run(node1, node2):', sess.run([node1, node2]))
print('sess.run(node3):', sess.run(node3))

sess.run(node1, node2): [3.0, 4.0]
sess.run(node3): 7.0


## Placeholder

- graph를 미리 만들어 놓고 실행하는 단계에서 값을 전달하려면?
- Placeholder라는 node를 만들고 run할 때 값을 전달한다.

In [10]:
a = tf.placeholder(tf.float32)
b = tf.placeholder(tf.float32)
adder_node = a + b # + provides a shortcut for tf.add(a, b)

In [11]:
print(sess.run(adder_node, feed_dict={a: 3, b: 4.5}))
print(sess.run(adder_node, feed_dict={a:[1, 3], b: [2, 4]}))

7.5
[3. 7.]


## Everything is Tensor

In [12]:
3 # a rank 0 tensor; this is a scalar with shape []
[1, 2, 3] # a rank 1 tensor; this is a vector with shape [3]
[[1, 2, 3], [4, 5, 6]] # a rank 2 tensor; a matrix with shape [2, 3]
[[[1, 2, 3]], [[7, 8, 9]]] # a rank 3 tensor with shape [2, 1, 3]

[[[1, 2, 3]], [[7, 8, 9]]]

## Tensor Ranks, Shapes and Types

- Rank 0, Scalar, s = 483
- Rank 1, Vector(magnitude and direction), v = [1.1, 2.2, 3.3]
- Rank 2, Matrix(table of numbers) m = [[1, 2, 3], [4, 5, 6]]
- Rank 3, 3-Tensor(cube of numbers) t = [[[2], [3], [4]], [[2], [3], [4]], [[2], [3], [4]]]
- Rank n, n-Tensor

- Shape [], Dimension number 0-D, Scalar
- Shape [D0], Dimension number 1-D, [5]
- Shape [D0, D1], 2-D, [3, 4]
- Shape [D0, D1, D2], 3-D, [3, 4, 5]
- Shape [D0, ..., Dn-1], n-D, [3, ... n-1]

- Data type, Python type
- DT_FLOAT, tf.float32
- DT_DOUBLE, tf.float64
- DT_INT8, tf.int8
- DT_INT32, tf.int32

## TensorFlow Machanics

- 1.Build graph using TensorFlow operations
- 2.feed data and run graph(operation)
    - sess.run(op, feed_dict={x: x_data})
- 3.update variables in the graph (and return values)